llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -168,11 +168,21 @@ class AMDGPUInformationCache : public InformationCache {
     return ST.supportsGetDoorbellID();
   }
 
-  std::pair<unsigned, unsigned> getFlatWorkGroupSizes(const Function &F) {
+  std::optional<std::pair<unsigned, unsigned>>
+  getFlatWorkGroupSizeAttr(const Function &F) const {
+    Attribute A = F.getFnAttribute("amdgpu-flat-work-group-size");
+    if (!A.isStringAttribute())
```

Comment:
This amounts to querying the attribute twice. If you're going to apply your own maximum clamping later, can you directly parse the attribute instead of going through the smarter wrapper applying the clamp? i.e. use getIntegerPairAttribute? This would be easier to follow and you can avoid the second map lookup